File: python/crypto_api/elgamal/elgamal_encryption.py
from crypto_api.elgamal.elgamal_data_classes import *
from crypto_api.math_lib.multiplicative_group import MultiplicativeGroup


class ElGamalEncryption:
    @staticmethod
    def key_gen(prime: int, generator: int) -> ElGamalKeyPair:
        group: MultiplicativeGroup = MultiplicativeGroup(prime)
        secret_key: int = group.random_element()
        public_key: int = group.pow_mod(generator, secret_key)

        return ElGamalKeyPair(secret_key, public_key, ElGamalKeyParams(prime, generator))

    @staticmethod
    def enc(other_public_key: int, key_params: ElGamalKeyParams, message: int) -> ElGamalCipherText:
        group: MultiplicativeGroup = MultiplicativeGroup(key_params.prime)

        y: int = group.random_element()
        gy: int = group.pow_mod(key_params.generator, y)

        c2: int = group.mul_mod(
            group.pow_mod(other_public_key, y),
            message
        )

        c1 = gy
        return ElGamalCipherText(c1, c2)

    @staticmethod
    def dec(secrect_key: int, key_params: ElGamalKeyParams, cipherText: ElGamalCipherText) -> int:
        group: MultiplicativeGroup = MultiplicativeGroup(key_params.prime)

        shared_key = group.pow_mod(cipherText.c1, secrect_key)
        inverse_shared_key = group.mul_invert_mod(shared_key)

        return group.mul_mod(cipherText.c2, inverse_shared_key)

File: java/cryptoapi/elgamal/ElGamalEncryption.java
package cryptoapi.elgamal;

import cryptoapi.math_lib.CyclicMultiplicativeGroup;
import cryptoapi.math_lib.MathMisc;

import java.math.BigInteger;

public class ElGamalEncryption {

    public static ElGamalKeyPair keyGen(int secPar) {
        CyclicMultiplicativeGroup group = new CyclicMultiplicativeGroup(MathMisc.randomBigIntPrime(secPar));

        ElGamalKeyParams publicParams = new ElGamalKeyParams(group.randomElement(), group.m);
        BigInteger secretKey = group.randomElement();
        BigInteger publicKey = group.powMod(publicParams.generator, secretKey);

        return new ElGamalKeyPair(publicParams, secretKey, publicKey);
    }

    public static ElGamalCipherText enc(BigInteger otherPublicKey, ElGamalKeyParams keyParams, BigInteger message) {
        CyclicMultiplicativeGroup group = new CyclicMultiplicativeGroup(keyParams.prime);

        BigInteger y = group.randomElement();
        BigInteger c1 = group.powMod(keyParams.generator, y);

        BigInteger c2Prime = group.powMod(otherPublicKey, y);
        BigInteger c2 = group.mulMod(c2Prime, message);

        return new ElGamalCipherText(c1, c2);
    }

    public static BigInteger dec(ElGamalKeyPair keypair, ElGamalCipherText cipherText) {
        CyclicMultiplicativeGroup group = new CyclicMultiplicativeGroup(keypair.keyParams.prime);

        // Reconstructing Shared key
        BigInteger k = group.powMod(cipherText.c1, keypair.secretKey);
        BigInteger kinv = group.mulInvertMod(k);

        return group.mulMod(cipherText.c2, kinv);
    }
}
